ci-matrix.py

```diff
@@ -117,22 +120,50 @@ def generate_matrix_entries(
     return matrix_entries
 
 
+def find_runner(runners: dict[str, Any], platform: str, arch: str) -> str:
+    # Find a matching platform first
+    match_platform = [
+        runner for runner in runners if runners[runner]["platform"] == platform
+    ]
+
+    # Then, find a matching architecture
+    match_arch = [
+        runner for runner in match_platform if runners[runner]["arch"] == arch
+    ]
+
+    # If there's a matching architecture, use that
+    if match_arch:
+        return match_arch[0]
+
+    # Otherwise, use the first with a matching platform
+    if match_platform:
+        return match_platform[0]
+
+    raise RuntimeError(f"No runner found for platform {platform!r} and arch {arch!r}")
+
+
 def add_matrix_entries_for_config(
     matrix_entries: list[dict[str, str]],
     target_triple: str,
     config: dict[str, Any],
     platform: str,
+    runners: dict[str, Any],
     directives: set[str],
 ) -> None:
     python_versions = config["python_versions"]
     build_options = config["build_options"]
     arch = config["arch"]
+    runner = find_runner(runners, platform, arch)
 
     # Create base entry that will be used for all variants
     base_entry = {
         "arch": arch,
         "target_triple": target_triple,
         "platform": platform,
+        "runner": runner,
+        # If `run` is in the config, use that — otherwise, default to if the
+        # runner architecture matches the build architecture
+        "run": str(config.get("run", runners[runner]["arch"] == arch)).lower(),
```

Comment:
Once we do #431, we might not have access to the runners needed for `run`. We'll have to do something a bit more complicated for PGO targets in that case but I don't handle this here.